File: crates/trusty-agents/ui/src/lib/chatHistory.ts
// Chat rehydration from the durable persona log (#4278).
//
// Why: `messages` (`stores/app.ts`) starts as an empty Map on every load and
// is only ever appended to by `InputArea`, `eventBridge`, and `Sidebar`.
// Nothing read the persisted log back, so a plain page reload discarded the
// whole visible conversation while the durable record sat untouched in
// trusty-memory — a data-loss bug, not a missing feature. The two surfaces
// that look like they could serve cannot: `GET /api/tasks` returns
// `PmResponse` envelopes carrying NO request text (see `taskHistory.ts`), so
// it can never restore a prompt; and `GET /api/workstreams/:name/history` is
// the 10-item digest `resumeWorkstream` already injects as a banner, which the
// owner ruled does not satisfy this issue. `GET /api/agents/:name/chat-history`
// reads the `chat_session` projection keyed `persona-{agent}` — the only store
// holding both halves of a turn.
//
// What: `fetchChatHistory` — the thin REST wrapper, returning an unavailable
// envelope rather than throwing so a down daemon can never break the chat.
// `historyToMessages` — the pure, unit-tested mapper from the wire's flat
// `{role, content}` list to the `Message` shape `ChatView` renders.
//
// Bounded by construction: the `persona-{agent}` session is continuous and
// never rolls over, so the initial load asks for `DEFAULT_HISTORY_LIMIT`
// messages and `has_more` + the `before` cursor drive lazy loading of older
// turns on demand.
//
// Test: `chatHistory.test.ts`.

import { get } from 'svelte/store';

import { apiBase } from './api-config';
import {
  canLoadOlderChat,
  conversationKey,
  activeConversationKey,
  messages,
  isRunning,
  type ChatHistoryCursor,
  chatHistoryCursor,
  getCurrentApiToken,
  hydrateMessages,
  loadingOlderChat,
  prependMessages,
  type Message,
} from '../stores/app';

/** One persisted message, as `trusty_common`'s `ChatMessage` serializes. */
export interface ChatHistoryMessage {
  role: string;
  content: string;
}

/** `GET /api/agents/:name/chat-history` response body. */
export interface ChatHistoryPage {
  /**
   * False whenever there is nothing to rehydrate — the agent binds no memory
   * palace, its session does not exist yet, or trusty-memory is unreachable.
   * The route answers 200 in all three cases, so the chat view renders an
   * empty conversation rather than an error.
   */
  available: boolean;
  session_absent?: boolean;
  messages: ChatHistoryMessage[];
  /**
   * ABSOLUTE index of this page's first message. Passed straight back as the
   * next `until` to page backwards — see `fetchChatHistory`.
   */
  start: number;
  /** Total messages in the session, not in this page. */
  total: number;
  /** Whether older messages exist before this page's oldest. */
  has_more: boolean;
  /** Session-level timestamp; the store keeps no per-message time. */
  updated_at: string | null;
  reason?: string;
}

/**
 * Initial page size, in MESSAGES (a turn is a user/assistant pair, so this is
 * roughly 50 turns). Named in messages because that is what the store holds.
 */
export const DEFAULT_HISTORY_LIMIT = 100;

/**
 * The shape every failure path returns, so callers branch only on `available`.
 * Built per call rather than shared, so a `reason` can be attached without
 * mutating a module-level object every later caller would then see.
 */
function emptyPage(reason: string): ChatHistoryPage {
  return {
    available: false,
    messages: [],
    start: 0,
    total: 0,
    has_more: false,
    updated_at: null,
    reason,
  };
}

function authHeaders(): Record<string, string> {
  const token = getCurrentApiToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

/**
 * Why: rehydration runs on app start, when the sidecar may still be binding
 * its socket. A throw here would surface as an unhandled rejection during
 * bootstrap and leave the chat blank anyway, so every failure degrades to the
 * same empty page the backend returns for "nothing persisted yet".
 * What: `GET /api/agents/:name/chat-history?limit=&until=`. `until` is an
 * ABSOLUTE exclusive end index, omitted for the newest page. Absolute indexing
 * is what makes paging backwards stable while new turns append — an
 * offset-from-the-end cursor shifts by the number of appends and replays them.
 * Test: `fetchChatHistory_returns_empty_page_on_http_error`,
 * `fetchChatHistory_returns_empty_page_on_network_failure`,
 * `fetchChatHistory_requests_the_bounded_window`,
 * `fetchChatHistory_sends_until_only_when_paging_back`.
 */
export async function fetchChatHistory(
  agent: string,
  limit = DEFAULT_HISTORY_LIMIT,
  until?: number,
): Promise<ChatHistoryPage> {
  const cursor = until === undefined ? '' : `&until=${until}`;
  try {
    const r = await fetch(
      `${apiBase()}/api/agents/${encodeURIComponent(agent)}/chat-history` +
        `?limit=${limit}${cursor}`,
      { headers: authHeaders() },
    );
    if (!r.ok) return emptyPage(`chat history request failed (HTTP ${r.status})`);
    const body = (await r.json()) as Partial<ChatHistoryPage>;
    return {
      available: body.available === true,
      session_absent: body.session_absent === true,
      messages: Array.isArray(body.messages) ? body.messages : [],
      start: typeof body.start === 'number' ? body.start : 0,
      total: typeof body.total === 'number' ? body.total : 0,
      has_more: body.has_more === true,
      updated_at: typeof body.updated_at === 'string' ? body.updated_at : null,
      reason: body.reason,
    };
  } catch (e) {
    return emptyPage(`chat history unreachable: ${e}`);
  }
}

/**
 * Roles the chat store renders. `chat_turn_append` only ever writes `user` and
 * `assistant`, but the store's history blob is untyped JSON, so an unknown
 * role is mapped to `system` rather than cast blindly into a union it may not
 * belong to.
 */
function toRole(role: string): Message['role'] {
  return role === 'user' || role === 'assistant' ? role : 'system';
}

/**
 * Why: `ChatView` renders `Message`, not the wire shape, and it stamps a
 * timestamp on every bubble. The persisted store carries NO per-message
 * timestamp — `ChatMessage` is `{role, content}` — so a per-turn time would
 * have to be invented. It is not: every rehydrated bubble takes the session's
 * `updated_at`, the only real time the backend has, and the caller passes
 * `now` only as the fallback for a session that reports none.
 * What: pure map from the wire list to `Message[]`, preserving order.
 * `speaker` is stamped on assistant bubbles so a rehydrated conversation is
 * attributed the same way a live one is.
 * Test: `historyToMessages_maps_roles_and_preserves_order`,
 * `historyToMessages_stamps_session_time_not_a_fabricated_one`,
 * `historyToMessages_labels_unknown_roles_as_system`.
 */
export function historyToMessages(
  page: ChatHistoryPage,
  speaker: string,
  now: number,
): Message[] {
  const ts = page.updated_at ? Date.parse(page.updated_at) : NaN;
  const timestamp = Number.isNaN(ts) ? now : ts;
  return page.messages.map((m, i) => {
    let role = toRole(m.role);
    let content = m.content ?? '';
    let activity: Partial<Message> = {};
    if (m.role === 'system') {
      try {
        const event = JSON.parse(content);
        if (event?.kind === 'trusty.listener-event' && event.version === 1 && typeof event.listener === 'string' && typeof event.event_type === 'string') {
          role = 'event';
          if (typeof event.event_id === 'string') activity.eventId=event.event_id;
          content = [event.listener, event.event_type, typeof event.from === 'string' ? event.from : '', typeof event.subject === 'string' ? event.subject : ''].filter(Boolean).join(' · ');
        } else if (event?.kind === 'trusty.tool-activity' && event.version === 1 && typeof event.tool === 'string' && typeof event.call_id === 'string' && ['running','complete','error'].includes(event.status)) {
          role='tool'; content=''; activity={toolName:event.tool.slice(0,160),toolCallId:event.call_id,activityStatus:event.status === 'running' ? 'error' : event.status};
        }
      } catch { /* Ordinary system content stays an ordinary banner. */ }
    }
    return {
      // Stable and collision-free against live ids, which are uuid/task-based.
      id: `history-${page.start + i}`,
      role,
      content,
      timestamp,
      ...activity,
      ...(role === 'assistant' ? { speaker } : {}),
    };
  });
}

/**
 * Why: the persona log is keyed by AGENT (`persona-{agent}`) while the chat
 * store is keyed by project, so rehydration has to answer "whose history, and
 * under whose name" before it can fetch. `activeAgentId` is `null` for the
 * base ctrl/PM session (see its doc comment in `stores/app.ts`), and the
 * roster can still be empty on a cold start when the catalog fetch has not
 * landed — both have to resolve to something rather than fetching
 * `persona-null` or stamping `undefined` on every restored bubble.
 * What: pure resolution of `{agentId, speaker}` from the selection and the
 * roster. Extracted so the mapping is unit-testable, leaving `App.svelte` with
 * a one-line call rather than untested glue.
 * Test: `resolveRehydrationTarget_*`.
 */
export function resolveRehydrationTarget(
  activeAgentId: string | null,
  roster: { id: string; label: string }[],
): { agentId: string; speaker: string } {
  const agentId = activeAgentId ?? 'ctrl';
  const speaker = roster.find((entry) => entry.id === agentId)?.label ?? 'Assistant';
  return { agentId, speaker };
}

export interface RehydrateResult {
  /** Messages seeded into the store. 0 when nothing was persisted. */
  seeded: number;
  /** Whether older messages remain, for the "load earlier" affordance. */
  hasMore: boolean;
  /**
   * Why the load produced nothing, when it produced nothing. Every
   * `available: false` path — no palace bound, no session yet, daemon down,
   * malformed payload — arrives here, and all four render as an identically
   * empty chat. Without this the caller cannot tell a first-run empty
   * conversation from a broken one, so a real failure is invisible.
   */
  reason?: string;
}

/**
 * Why: the whole point of #4278 — on load, put the persisted conversation back
 * into the chat view. Kept here rather than inline in `App.svelte` so the path
 * is unit-testable against the real store with a stubbed `fetch`, instead of
 * being untestable component glue.
 * What: fetches the newest bounded page for `agentId` and seeds it into
 * `projectId`'s bucket via `hydrateMessages`, which refuses to clobber
 * anything already there. Returns what it did so a caller can decide whether
 * to offer "load earlier".
 * Test: `rehydrateChat_restores_the_persisted_conversation`,
 * `rehydrateChat_is_a_noop_when_nothing_is_persisted`,
 * `rehydrateChat_does_not_clobber_a_message_typed_during_bootstrap`.
 */
const historyRequests = new Map<string, number>();
const historyEnds = new Map<string, number>();
const savedCursors = new Map<string, { anchor: Message; cursor: ChatHistoryCursor }>();
function rememberCursor(key: string, cursor: ChatHistoryCursor) {
  const anchor = get(messages).get(key)?.[0];
  if (anchor) savedCursors.set(key, { anchor, cursor });
}
function cachedCursor(key: string): ChatHistoryCursor | null {
  const saved = savedCursors.get(key);
  return saved && get(messages).get(key)?.includes(saved.anchor) ? saved.cursor : null;
}

export async function rehydrateChat(
  agentId: string,
  speaker: string,
  projectId: string,
  limit = DEFAULT_HISTORY_LIMIT,
): Promise<RehydrateResult> {
  const key = conversationKey(projectId, agentId);
  const generation = (historyRequests.get(key) ?? 0) + 1;
  historyRequests.set(key, generation);
  const selected = () => get(activeConversationKey) === key;
  const cached = cachedCursor(key);
  if (cached) {
    if (selected()) chatHistoryCursor.set(cached);
    return { seeded: 0, hasMore: cached.hasMore };
  }
  if (selected()) chatHistoryCursor.set(null);
  const page = await fetchChatHistory(agentId, limit);
  if (historyRequests.get(key) !== generation) return { seeded: 0, hasMore: cachedCursor(key)?.hasMore ?? false };
  if (page.available) historyEnds.set(key,page.start+page.messages.length);
  else if(page.session_absent)historyEnds.set(key,0);
  if (!page.available || page.messages.length === 0) {
    if (selected()) chatHistoryCursor.set(null);
    return { seeded: 0, hasMore: false, reason: page.reason };
  }
  const restored = historyToMessages(page, speaker, Date.now());
  const seeded = hydrateMessages(key, restored);
  if (seeded) rememberCursor(key, { agentId, speaker, projectId, start: page.start, hasMore: page.has_more });
  const cursor = cachedCursor(key);
  if (selected()) chatHistoryCursor.set(cursor);
  return { seeded: seeded ? restored.length : 0, hasMore: cursor?.hasMore ?? false };
}

/**
 * Why: the owner's volume requirement pairs a bounded initial load with lazy
 * loading of older turns on demand — without this, a bounded load is just
 * truncation. Driven by `ChatView`'s "Load earlier messages" control, which
 * renders off `canLoadOlderChat`.
 * What: takes no target — the cursor names the agent, the speaker, and the
 * bucket, so this cannot page one conversation into another's view. Gated on
 * `canLoadOlderChat`, so a cursor left over from a different agent or project
 * is refused here as well as hidden in the view. Fetches the page ending at the
 * cursor's absolute index, PREPENDS it so older turns land above what is
 * already rendered, and advances the cursor. Ids are namespaced by the page's
 * absolute `start`, so a prepended page cannot collide with the seeded one.
 * Test: `loadOlderChat_prepends_the_previous_page`,
 * `loadOlderChat_advances_the_cursor`,
 * `loadOlderChat_is_a_noop_without_a_cursor`,
 * `loadOlderChat_refuses_a_cursor_from_another_agent`,
 * `loadOlderChat_prepends_into_the_cursors_own_project`,
 * `loadOlderChat_disarms_the_cursor_at_the_start_of_history`.
 */
export async function loadOlderChat(
  limit = DEFAULT_HISTORY_LIMIT,
): Promise<RehydrateResult> {
  const cursor = get(chatHistoryCursor);
  if (!cursor || !get(canLoadOlderChat)) return { seeded: 0, hasMore: false };

  loadingOlderChat.set(true);
  try {
    const page = await fetchChatHistory(cursor.agentId, limit, cursor.start);
    if (!page.available || page.messages.length === 0) {
      const next = { ...cursor, hasMore: false };
      rememberCursor(conversationKey(cursor.projectId, cursor.agentId), next);
      if (get(chatHistoryCursor) === cursor) chatHistoryCursor.set(next);
      return { seeded: 0, hasMore: false, reason: page.reason };
    }
    const older = historyToMessages(page, cursor.speaker, Date.now());
    // The cursor's own bucket, never the currently-active one — the two can
    // differ, and prepending into the active one is how a restored conversation
    // leaks into an unrelated view.
    prependMessages(conversationKey(cursor.projectId, cursor.agentId), older);
    const next = { ...cursor, start: page.start, hasMore: page.has_more };
    rememberCursor(conversationKey(cursor.projectId, cursor.agentId), next);
    if (get(chatHistoryCursor) === cursor) chatHistoryCursor.set(next);
    return { seeded: older.length, hasMore: page.has_more };
  } finally {
    loadingOlderChat.set(false);
  }
}

/** Append completed incoming event turns without replacing live drafts or task messages. */
export async function refreshEventHistory(agentId:string,speaker:string,projectId:string):Promise<number> {
  const key=conversationKey(projectId,agentId);
  const end=historyEnds.get(key);
  const pages:ChatHistoryPage[]=[];
  let page=await fetchChatHistory(agentId,500);
  if(!page.available)throw new Error(page.reason??'Chat history is temporarily unavailable');
  const latestEnd=page.start+page.messages.length;
  if(end===undefined){
    if(get(isRunning))throw new Error('Incoming history refresh waits for the active task');
    const restored=historyToMessages(page,speaker,Date.now());
    const existingIds=new Set((get(messages).get(key)??[]).map(m=>m.id));
    const missing=restored.filter(m=>!existingIds.has(m.id));
    prependMessages(key,missing);
    historyEnds.set(key,latestEnd);
    const cursor={agentId,speaker,projectId,start:page.start,hasMore:page.has_more};
    rememberCursor(key,cursor);
    if(get(activeConversationKey)===key)chatHistoryCursor.set(cursor);
    return missing.length;
  }
  pages.push(page);
  while(page.start>end && page.has_more && pages.length<4){
    page=await fetchChatHistory(agentId,500,page.start);
    if(!page.available)throw new Error(page.reason??'Chat history is temporarily unavailable');
    pages.unshift(page);
  }
  if(page.start>end && page.has_more)throw new Error('Incoming history is too large to refresh safely');
  const restored=pages.flatMap(p=>historyToMessages(p,speaker,Date.now()));
  const groups:Message[][]=[];let group:Message[]=[];let safeEnd=latestEnd;
  const finish=(tail=false)=>{if(group[0]?.eventId){if(group.some(m=>m.role==='assistant'))groups.push(group);else if(tail)safeEnd=Math.min(safeEnd,Number(group[0].id.slice('history-'.length)));}group=[];};
  for(const message of restored){if(message.role==='event'){finish();group=[message];}else if(message.role==='user'){finish();}else if(group.length){group.push(message);}}
  finish(true);
  if(get(isRunning))throw new Error('Incoming history refresh waits for the active task');
  let added=0;
  messages.update(map=>{
    const list=map.get(key)??[];const next=[...list];const ids=new Set(next.map(m=>m.id));
    for(const rows of groups){const eventId=rows[0].eventId!;
      for(const row of rows){const index=Number(row.id.slice('history-'.length));if(index<end || ids.has(row.id))continue;
        next.push({...row,eventId});ids.add(row.id);added++;
      }
    }
    if(!added)return map;const updated=new Map(map);updated.set(key,next);return updated;
  });
  historyEnds.set(key,Math.max(historyEnds.get(key)??0,safeEnd));
  return added;
}

File: crates/trusty-agents/src/api/server/chat_history.rs
//! `GET /api/agents/:name/chat-history` — read back the durable persona chat
//! log so the GUI can rehydrate its chat view on reload (#4278).
//!
//! Why: every persona chat turn is already persisted — `spawn_persist_turn`
//! (`ctrl::pm_task::dispatch::persona_memory`) appends the user prompt and the
//! assistant response to a trusty-memory `chat_session` keyed
//! `persona-{agent}` (DOC-54 §9.1). Nothing could READ it back over HTTP, so
//! a plain page reload discarded the visible conversation while the durable
//! log sat untouched — a data-loss bug, not a missing feature. The two
//! surfaces that looked like they might serve are both wrong for this:
//! `GET /api/tasks` returns `PmResponse` envelopes that carry NO request text
//! (see `lib/taskHistory.ts`), so it can never restore a prompt; and
//! `GET /api/workstreams/:name/history` returns `ws:`-tagged drawers, which
//! the owner ruled a digest rather than the turn history this issue requires.
//! The `chat_session` projection is the only store holding both halves of a
//! turn, which is why the owner named it as the read source.
//!
//! What: resolves the agent's bound palace from `[[stores]].primary().palace`
//! (the same mapping `agent_kg` performs), dials `chat_session_recall` through
//! the `tools/call` envelope, and returns a BOUNDED slice of the flat
//! `Vec<ChatMessage>` history, newest-end first, with a cursor for fetching
//! older messages on demand. The bound is the owner's volume requirement: the
//! `persona-{agent}` session is continuous and never rolls over, so hydrating
//! every turn on launch does not scale.
//!
//! Honest limits, stated rather than papered over:
//!   - The bound is applied HERE, not upstream. `chat_session_recall` has no
//!     `limit`/`offset` parameter (`tools/chat_definitions.rs`), so the daemon
//!     still returns the whole session over the socket and this route slices
//!     it. That caps what the browser renders and parses, not what crosses the
//!     Unix socket; bounding the daemon read needs a trusty-memory change and
//!     is deliberately out of scope here.
//!   - `ChatMessage` is `{role, content}` with NO per-message timestamp
//!     (`trusty_common::memory_core::store::chat_sessions::types`), so the
//!     response carries only the session's `updated_at`. The client must not
//!     invent per-turn times it was never given.
//!   - Only turns that reached [`spawn_persist_turn`] are here. That call sits
//!     on the SUCCESS path in `persona.rs`, so an errored or cancelled turn is
//!     never appended, and a stream that never completed has no final response
//!     to append. `available: true` therefore means "this is the persisted
//!     log", never "this is every turn that ever happened" — the distinction
//!     the owner asked to have verified before trusting this source.
//!
//! An agent binding no palace persists nothing, so it has nothing to
//! rehydrate. That is an ordinary state, reported as `200` +
//! `available: false` + a reason (the `costs`/`agent_kg` precedent), never an
//! error the GUI has to render as a failure.
//!
//! Test: `super::tests::chat_history`.

use std::path::{Path, PathBuf};

use axum::{
    Json,
    extract::{Path as AxumPath, Query, State},
    http::StatusCode,
    response::{IntoResponse, Response},
};
use serde::Deserialize;
use serde_json::{Value, json};

use super::agent_kg::{describe_failure, parse_stores};
use super::agent_patch::resolve_agent_paths;
use super::state::AppState;
use crate::ctrl::pm_task::dispatch::persona_memory::session_id_for;
use crate::stores::status::PROBE_TIMEOUT;

/// Messages returned when the client names no `limit` — roughly 50 turns.
const DEFAULT_LIMIT: usize = 100;
/// Hard ceiling on one page, so a hand-crafted `limit` cannot ask the GUI to
/// render an unbounded session in one paint.
const MAX_LIMIT: usize = 500;

/// `?limit=<n>&before=<n>`.
///
/// `before` is the number of most-recent messages the client ALREADY holds —
/// an offset counted from the newest end, not a timestamp. A cursor counted
/// from the end is stable under appends: new turns land beyond the window the
/// client is paging backwards through, so lazy-loading older messages cannot
/// skip or duplicate one because a turn arrived mid-scroll.
#[derive(Debug, Default, Deserialize)]
pub(super) struct HistoryQuery {
    limit: Option<usize>,
    before: Option<usize>,
}

/// `GET /api/agents/:name/chat-history` — HTTP entry point.
pub(super) async fn agent_chat_history_route(
    State(_state): State<AppState>,
    AxumPath(name): AxumPath<String>,
    Query(q): Query<HistoryQuery>,
) -> Response {
    chat_history_at(
        &crate::agents::agents_dir_candidates(),
        &name,
        trusty_common::memory_rpc::resolve_memory_socket()
            .ok()
            .as_deref(),
        q.limit.unwrap_or(DEFAULT_LIMIT).clamp(1, MAX_LIMIT),
        q.before.unwrap_or(0),
    )
    .await
}

/// Core read against explicit agents dirs + a trusty-memory socket.
///
/// Why: the socket is injected so tests exercise every branch against a mock
/// daemon on a temp socket instead of the developer's live one — the
/// `agent_kg::kg_proxy_at` / `agent_stores::stores_at` pattern.
/// What: `400` on an invalid agent name, `404` on an unknown agent, `500` only
/// when the agent's own config cannot be read. Every other outcome — no palace
/// bound, daemon undiscoverable, daemon unreachable, session never created —
/// is a `200` carrying `available: false` and a reason, because each of those
/// means "nothing to rehydrate", which the chat view renders as an empty
/// conversation rather than an error.
/// Test: `chat_history_returns_bounded_newest_slice`,
/// `chat_history_before_cursor_pages_backwards`,
/// `chat_history_empty_when_no_palace_bound`,
/// `chat_history_empty_when_session_absent`,
/// `chat_history_degrades_when_memory_unreachable`,
/// `chat_history_unknown_agent_404`, `chat_history_rejects_traversal_name`.
pub(super) async fn chat_history_at(
    dirs: &[PathBuf],
    name: &str,
    memory_socket: Option<&Path>,
    limit: usize,
    before: usize,
) -> Response {
    if name.is_empty() || name.contains(['/', '\\']) || name == "." || name == ".." {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({ "error": "invalid agent name" })),
        )
            .into_response();
    }
    let Some((path, _package_dir)) = resolve_agent_paths(dirs, name) else {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({ "error": "unknown agent", "name": name })),
        )
            .into_response();
    };
    let raw = match tokio::fs::read_to_string(&path).await {
        Ok(s) => s,
        Err(e) => {
            tracing::warn!(?e, agent = name, path = %path.display(), "chat_history_at: read failed");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "failed to read agent config" })),
            )
                .into_response();
        }
    };

    let session_id = session_id_for(name);
    let (stores, config_error) = parse_stores(&raw);
    let Some(palace) = stores.primary().and_then(|b| b.palace.clone()) else {
        return unavailable(
            None,
            &session_id,
            "this agent binds no memory palace (`[[stores]].palace` is unset), so no chat turns are persisted",
            config_error,
        );
    };

    let session = match memory_socket {
        None => Err("trusty-memory daemon not discoverable (is it running?)".to_string()),
        Some(socket) => fetch_session(socket, &palace, &session_id).await,
    };
    match session {
        Err(reason) => unavailable(Some(palace), &session_id, &reason, config_error),
        Ok(session) => {
            let mut body = page(&palace, &session_id, &session, limit, before);
            if let Some(err) = config_error {
                body["config_error"] = Value::String(err);
            }
            (StatusCode::OK, Json(body)).into_response()
        }
    }
}

/// The `200` + `available: false` body every "nothing to rehydrate" path returns.
///
/// Shape-identical to the success body so a client branches only on
/// `available`, never on whether a field is present.
fn unavailable(
    palace: Option<String>,
    session_id: &str,
    reason: &str,
    config_error: Option<String>,
) -> Response {
    let mut body = json!({
        "available": false,
        "reason": reason,
        "palace": palace,
        "session_id": session_id,
        "messages": [],
        "total": 0,
        "has_more": false,
        "updated_at": Value::Null,
    });
    if let Some(err) = config_error {
        body["config_error"] = Value::String(err);
    }
    (StatusCode::OK, Json(body)).into_response()
}

/// Slice the newest `limit` messages ending `before` from the end.
///
/// Why: the owner's volume requirement — a continuous session that never rolls
/// over must not hydrate in full on launch. The window is
/// `[total - before - limit, total - before)`, clamped, so `before` walks
/// backwards one page at a time and `has_more` says whether anything older
/// remains.
/// What: pure over the decoded session value; a `history` that is absent or
/// not an array reads as an empty session rather than an error, because the
/// daemon's own contract already guarantees the field.
/// Test: `chat_history_returns_bounded_newest_slice`,
/// `chat_history_before_cursor_pages_backwards`.
fn page(palace: &str, session_id: &str, session: &Value, limit: usize, before: usize) -> Value {
    let history = session
        .get("history")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let total = history.len();
    let end = total.saturating_sub(before);
    let start = end.saturating_sub(limit);
    json!({
        "available": true,
        "palace": palace,
        "session_id": session_id,
        "messages": history[start..end].to_vec(),
        "total": total,
        "has_more": start > 0,
        "updated_at": session.get("updated_at").cloned().unwrap_or(Value::Null),
    })
}

/// Fetch the session, collapsing every failure into the reason a client renders.
///
/// Why: `chat_session_recall` is absent from trusty-memory's direct-dispatch
/// `TOOL_METHODS` allow-list, so it is only reachable wrapped in `tools/call`
/// — the same envelope `persona_memory::call_memory_tool` writes through. That
/// arm answers `result.content[0].text` as a JSON STRING, so the payload needs
/// decoding back out; it also reports EVERY tool failure as `-32603`, so a
/// session that simply does not exist yet cannot be told apart by code and is
/// matched on the daemon's own "session not found" wording instead.
/// What: `Ok(session)` when the daemon answers; `Err(reason)` for an absent
/// session, any other refusal, or a transport failure. Bounded by
/// [`PROBE_TIMEOUT`], the ceiling every cross-daemon read in this crate uses.
/// Test: `chat_history_empty_when_session_absent`,
/// `chat_history_degrades_when_memory_unreachable`.
async fn fetch_session(socket: &Path, palace: &str, session_id: &str) -> Result<Value, String> {
    let answer = trusty_common::memory_rpc::call_memory_tool_at_with_timeout(
        socket,
        "tools/call",
        json!({
            "name": "chat_session_recall",
            "arguments": { "palace": palace, "session_id": session_id },
        }),
        PROBE_TIMEOUT,
    )
    .await
    .map_err(|e| {
        let described = describe_failure(&e, palace);
        if described.contains("not found") || described.contains("does not exist") {
            format!("no persisted chat session `{session_id}` yet")
        } else {
            described
        }
    })?;

    let text = answer
        .pointer("/content/0/text")
        .and_then(Value::as_str)
        .ok_or_else(|| {
            "trusty-memory answered chat_session_recall without `content[0].text`".to_string()
        })?;
    serde_json::from_str::<Value>(text)
        .map_err(|e| format!("trusty-memory returned an undecodable chat session: {e}"))
}
